leader-election-impl/src/main/java/com/palantir/paxos/SqlitePaxosStateLogFactory.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.paxos;
+
+import java.sql.Connection;
+import java.util.List;
+import java.util.function.Supplier;
+
+import org.immutables.value.Value;
+import org.jdbi.v3.core.Jdbi;
+import org.jdbi.v3.sqlobject.SqlObjectPlugin;
+import org.jdbi.v3.sqlobject.customizer.Bind;
+import org.jdbi.v3.sqlobject.statement.SqlQuery;
+import org.jdbi.v3.sqlobject.statement.SqlUpdate;
+
+import com.palantir.common.persist.Persistable;
+
+public final class SqlitePaxosStateLogFactory<V extends Persistable & Versionable> {
+    public static final String NAMESPACE_TABLE = "_paxosNamespaces";
+
+    private final Supplier<Connection> connectionSupplier;
+    private final Jdbi jdbi;
+
+    private SqlitePaxosStateLogFactory(Supplier<Connection> connectionSupplier, Jdbi jdbi) {
+        this.connectionSupplier = connectionSupplier;
+        this.jdbi = jdbi;
+    }
+
+    public static <V extends Persistable & Versionable> SqlitePaxosStateLogFactory<V> create(
+            Supplier<Connection> connectionSupplier) {
+        Jdbi jdbi = Jdbi.create(connectionSupplier::get).installPlugin(new SqlObjectPlugin());
+        jdbi.withExtension(Queries.class, Queries::createNamespaceTable);
+        return new SqlitePaxosStateLogFactory<>(connectionSupplier, jdbi);
+    }
+
+    public PaxosStateLog<V> createPaxosStateLog(PaxosStateLogCreationContext context) {
+        jdbi.withExtension(Queries.class, queries -> queries.registerNamespace(context.atlasNamespace()));
+        return SqlitePaxosStateLog.create(context.physicalTableName(), connectionSupplier);
+    }
+
+    public List<String> getAllRegisteredNamespaces() {
+        return jdbi.withExtension(Queries.class, Queries::getAllRegisteredNamespaces);
+    }
+
+    @Value.Immutable
+    public interface PaxosStateLogCreationContext {
+        String atlasNamespace();
+        String physicalTableName();
```

Comment:
How is this going to be generated?

Reply:
please can we just have the physical table name be a column? As in, one table for all namespaces

Reply:
and then the other question is something like: does

`SELECT DISTINCT(namespace) AS namespace FROM log` give us what we need?

Reply:
The generation was meant to happen inside `timelock-impl` or `TimeLockAgent`, we'd write a function `(PaxosUseCase, Client, PaxosFunction) -> String` (where `PaxosFunction` is acceptor or learner).

@j-baker We benchmarked that approach against separate tables, and we found that the separate tables gave us (marginal but reproducibly) better write perf as well as (~30%) less disk usage, which is why we went for it. 

That said it looks like there are complexity costs here (SELECT DISTINCT needs benchmarking, and should hopefully be fast since there would be a primary key index, plus it is what we want) though.

Reply:
The suggested solution is cleaner, but I prefer this approach due to the perf benefits